### ISSUE TITLE
Watchman hangs for ~2m when passed an inaccessible sockname path

### ISSUE THREAD

**Issue by andrew-stripe** [facebook/watchman#543]
On 4.9.0, running `watchman get-sockname --sockname /some/inaccessible/path` hangs for ~2 minutes. 
```
$ time watchman get-sockname --sockname /foo/bar
2017-10-26T21:28:28,701: [cli] unable to talk to your watchman on /foo/bar! (No such file or directory)

real	2m1.620s
user	0m0.001s
sys	0m0.184s
```

I believe this occurs because it attempts to start a new Watchman daemon but that daemon fails because it can't write to the directory. The client gets an `ENOENT` and loops: https://github.com/facebook/watchman/commit/155d656ca218ca8d677638b8cdd3d9196d68a2d2.

One fix would be replacing `should_start(errno)` with `errno == ECONNREFUSED` but I'm not sure if there are other implications of that change.

**Comment by wez** [comment]
Do you see a reasonable error message in the watchman log file for the server?

What's problematic here, resolution wise, is that the client doesn't know why the server isn't starting up, and we can't ask it because we can't talk to it.  We also don't know for sure whether we could start it up without a TOCTOU prone check to guess whether things would work out.

Is the main aspect of this for you the duration of the timeout, or the discoverability of the issue?

Looking a bit deeper, we have code that sanity checks ownership on the dirs, but I think it only kicks in when you don't override it via the command line options:

https://github.com/facebook/watchman/blob/master/main.cpp#L571

If you're feeling motivated, do you feel up to making a PR that factors out the ownership check so that it can be applied to the sockname path passed in directly via the CLI?  Even just making a rough cut at that to see if it would detect the problem would help reach a resolution.

**Comment by andrew-stripe** [comment]
You're right, this is only an issue when passing an explicit `--sockname`. I'm reworking our setup to use the default paths for unrelated reasons so this won't be a problem for me in practice going forward.

I didn't see any server log message because I had also configured a `--logfile`  in the same non-existent directory as the socket. When I point to a valid logfile location, the server logs do show `2017-10-26T21:18:44,199: [listener] bind(/foo/bar): No such file or directory`.

It wasn't until I was preparing to file this issue that I discovered `get-sockname` would still time out after 2 minutes. In my opinion https://github.com/facebook/watchman/commit/155d656ca218ca8d677638b8cdd3d9196d68a2d2 introduces too long a total timeout without any logging. 

It seems like the retry logic introduced in that commit is duplicative of [the retry loop in `w_stm_connect_unix`](https://github.com/facebook/watchman/blob/8f9018246c2a332756e7adefa8c668086a03c728/stream_unix.cpp#L250-L255). Why have separate retry loops in both places?
